compiler/src/iree/compiler/Codegen/Common/GPU/GPULowerToGlobalLoads.cpp

```diff
@@ -0,0 +1,338 @@
+        delinearizeIndex(linearizedGatherIndices, sourceType.getShape());
+    auto linearizedBaseIndices =
+        getSubgroupStoreBaseIndex(subgroupId, inductionVar);
+    auto delinearizedLocalIndices =
+        delinearizeIndex(linearizedBaseIndices, localType.getShape());
+    rewriter.create<IREE::GPU::GlobalLoadDMAOp>(
+        loc, copy.getOperand(0), delinearizedGlobalIndices,
+        copy.getOutputs()[0], delinearizedLocalIndices);
+  }
+
+  copy->erase();
+  return true;
+}
+
+} // namespace
+
+static bool checkEligibilityForGlobalLoadDMA(linalg::CopyOp copy) {
+  // source must be global address and target must be workgroup address.
+  auto sourceType = cast<MemRefType>(copy.getOperand(0).getType());
+  auto targetType = cast<MemRefType>(copy.getOutputs().front().getType());
+  if (!hasGlobalMemoryAddressSpace(sourceType)) {
+    LDBG("-- Op: "
+         << *copy
+         << "\n-- has source memory address space other than global.\n");
+    return false;
+  }
+  if (targetType.getMemorySpace() !=
+      gpu::AddressSpaceAttr::get(copy->getContext(),
+                                 gpu::GPUDialect::getWorkgroupAddressSpace())) {
+    LDBG("-- Op: "
+         << *copy
+         << "\n-- has target memory address space other than workgroup.\n");
+    return false;
+  }
+  // TODO: check that the copy's target memref is not a subview.
+  return true;
+}
+
+namespace {
+struct GPULowerToGlobalLoadsPass final
+    : impl::GPULowerToGlobalLoadsPassBase<GPULowerToGlobalLoadsPass> {
+
+  void runOnOperation() override {
+    MLIRContext *context = &getContext();
+    auto funcOp = getOperation();
+
+    // we will do this before/at the time of tiling, we need:
+    // 1. tiling configurations, to compute how many to generate.
+
+    SmallVector<linalg::CopyOp> copies;
+    funcOp->walk<WalkOrder::PreOrder>([&](Operation *op) {
+      // Walk in PreOrder so that parent operations are visited before children,
+      // thus allowing all operations contained within thread/warp/lane foralls
+      // to be skipped.
+      if (auto forallOp = dyn_cast<scf::ForallOp>(op)) {
+        // Skip ops contained within forall ops with thread/warp/lane mappings.
+        if (forallOpHasMappingType<IREE::GPU::LaneIdAttr,
+                                   gpu::GPUWarpMappingAttr,
+                                   gpu::GPUThreadMappingAttr>(forallOp)) {
+          return WalkResult::skip();
+        }
+      }
+      if (auto copy = dyn_cast<linalg::CopyOp>(op)) {
+        if (checkEligibilityForGlobalLoadDMA(copy)) {
+          copies.push_back(copy);
+        } else {
+          LDBG("Skipping copy op: "
+               << *copy
+               << " because it is not eligible for global load DMA.\n");
+        }
+      }
+      return WalkResult::advance();
+    });
+
+    std::optional<SmallVector<int64_t>> workgroupSize =
+        mlir::iree_compiler::getWorkgroupSize(funcOp);
+    if (!workgroupSize) {
+      funcOp.emitOpError(
```

Comment:
This'll emit errors even if we wouldn't've run the pass in the first place (ex. because there aren't any DMA copies to lower to)?

Reply:
To avoid this, I changed to only enable this pass when the direct-load option is turned on.